### ISSUE TITLE
Problem with Newley released types

### ISSUE THREAD

**Issue by KevinGruber** [remy/nodemon#2206]
- Versions: node@v22.1.0, darwin@23.2.0
- `nodemon -v`: 3.1.1
- Operating system/terminal environment (powershell, gitshell, etc):  MacOs, terminal
- Command you ran: Via api in typescript

### Expected behaviour
When importing nodemon there is a default export for functionality defined in the types.

### Actual behaviour
With the addition of https://github.com/remy/nodemon/pull/2204 there where types added to the project, but no default export types for functionality -> that leads to a not usable nodemon in typescript

### Steps to reproduce
import nodemon from 'nodemon';

->  nodemon is not a module

---

**Comment by remy** [comment]
Just checking, but does require work? And can you import using `import * as nodemon from 'nodemon'?

**Comment by KevinGruber** [comment]
const nodemon = require('nodemon'); - WORKS doesn't load types

import * as nodemon from 'nodemon'; - DOESN't work

I am pretty sure it is because there is no export of the functionality of nodemon.

Based on the types nothing is exported

**Comment by remy** [comment]
I need to have a play. I only added types as an attempt to help people who were requiring the module, but I don't personally use TypeScript ([some context](https://remysharp.com/2024/02/23/why-my-code-isnt-in-typescript)). If you have any advice, please do share, otherwise I'll tinker this week 👍

**Comment by remy** [comment]
Okay, to clarify, `import nodemon from 'nodemon'` does work, i.e. this code works:

```
nodemon({}).on('start', () => {
  console.log('started');
});
```

But the types aren't being found by tools like VS Code.

This is my code:

package.json

```json
{
  "name": "2206",
  "version": "1.0.0",
  "main": "index.js",
  "dependencies": {
    "nodemon": "^3.1.1"
  }
}
```

index.js (the sub process):

```js
console.log('Hello, World!');
```

And the run script: test.mjs

```js
import nodemon from 'nodemon';

nodemon({}).on('start', () => {
  console.log('started');
});
```

From the command line:

<img width="687" alt="SCR-20240528-jwdv" src="https://github.com/remy/nodemon/assets/13700/6fe4caf9-531a-4581-8c96-8073d5991a65">

@KevinGruber you're saying it's undefined, can you share your code?

**Comment by KevinGruber** [comment]
Ok sry for the misunderstanding.

The runtime code works yes, but the typescript doesn't build because of "wrong" types.

undefined in the sense of types in TS

Beforehand it worked because there where no types (and I configured ts to allow for no types).

If you want I can make a PR with adjusted types?

**Comment by remy** [comment]
Yeah, if you could do a PR I'd appreciate it. I don't want to convert the project to typescript, but I do want to expose the types to help devs (which is why I've just added index.d.ts)

**Comment by KevinGruber** [comment]
I have the types already working, I now ran into the problem that they are not fitting the api approach types.

Can you help me and send me a list of allowed attributes? Like script etc.

are those than just merged with the current NodemonSettings? Or is NodemonSettings just the json config representation?

**Comment by remy** [comment]
This:

> NodemonSettings just the json config representation

Sod, there's more props, I'll take a look and see if I can get a complete list. Annoyingly the `--dump` command doesn't quite list it out, because the inbound config is rearranged to the internal config (like `script` is moved to `execOptions`.

**Comment by remy** [comment]
If you have a branch, send up a work in progress PR and I can add to it (I think…)

**Comment by remy** [comment]
Found the following - pretty sure that's it.

```ts
script?: string; 
ext?: string; // "js,mjs" etc (should really support an array of strings, but I don't think it does right now)
events?: { [key: string]: string }; 
env?: { [key: string]: string };
exec?: string; // node, python, etc
execArgs?: string[]; // args passed to node, etc
```

**Comment by arpowers** [comment]
@remy the types still aren't correct in latest version, type errors happen when using the documented API (the API that was 100% working prior to 3.1.1.

**Comment by KevinGruber** [comment]
@arpowers what do you mean with working before, it wasn't using any types before

Can you show me the error?

**Comment by remy** [comment]
@arpowers a screenshot would be useful here, or othewise some code for context. Ta.

**Comment by arpowers** [comment]
Sure here are the  errors that show up after upgrade: 

Let me know if you want a link to the code.

```bash 
@fiction/core/plugin-env/restart.ts:19:13 - error TS2503: Cannot find namespace 'nodemon'.

19     config: nodemon.Settings
               ~~~~~~~

@fiction/core/plugin-env/restart.ts:65:8 - error TS2339: Property 'on' does not exist on type '(settings: NodemonSettings) => Nodemon'.

65       .on('log', () => {})
          ~~
```

<img width="696" alt="Screenshot 2024-05-29 at 1 37 13 PM" src="https://github.com/remy/nodemon/assets/308134/2ef43065-629c-4ccc-9b71-5640811d4c07">

**Comment by arpowers** [comment]
I should mention that I'm using `"@types/nodemon": "^1.19.6",` as well, let me try and  uninstall that and see if anything changes.

**Comment by remy** [comment]
Oooh. I know what it is. Though, not sure how you represent it in Typescript. It's because the `nodemon` function returns itself, and it's supposed to be a singleton. 

This isn't a fix, but if you chain the. `on` to the call to nodemon, I suspect it will work.

I'll see if I can fix the types in the morning, or if anyone gets a PR before then, happy to merge.

Unrelated: I wonder who is maintaining @types/nodemon...that isn't mine... 🤔

**Comment by arpowers** [comment]
no hurry, just locked version for now.

**Comment by remy** [comment]
@KevinGruber or @arpowers I'd appreciate some eyes on my change: https://github.com/remy/nodemon/pull/2208/files - I've tested it and it appears to work (with autocomplete), but more importantly the code actually still runs. 

Again, I'm not a regular TS dev, so wary it could be wrong.

**Comment by KevinGruber** [comment]
@remy yes that works for me I tested both use cases

I wasn't aware that you can use nodemon as a singleton like that. Sry for missing that :-(

I noticed another thing yesterday can you provide string and string[] to nodeArgs and execArgs?

I took your suggestion and it took only string[] but my old code provided string.

If it supports both and is expected I would change the type to string | string[]

**Comment by arpowers** [comment]
@remy all working, it caused some type errors, but only because my code was wrong. Working exactly as it should. Thanks!

**Comment by kpman** [comment]
> Unrelated: I wonder who is maintaining @types/nodemon...that isn't mine... 🤔

@remy The `@types/nodemon` author's information:
```
Type definitions for nodemon 1.19
Project: http://nodemon.io
Definitions by: Emily Marigold Klassen <https://github.com/forivall>
Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
```

Maybe you could contact it or just portint the type definitions into this repo.
The type definitions in 3.1.3 are not worked when using as the module.

**Comment by remy** [comment]
@kpman can you clarify this (or provide a working example)?

>  The type definitions in 3.1.3 are not worked when using as the module.

**Comment by GabenGar** [comment]
@remy 
Just FYI you can use [`arethetypeswrong`](https://arethetypeswrong.github.io/?p=nodemon%403.1.4) to check if the types are declared properly.

**Comment by remy** [comment]
@GabenGar that just tells me what you told me, that they're wrong. I'm not a typescript person, and when I do use it (for work) it's a day of deciphering ts error messages.

So if you can (I'd hugely appreciate it) explain in plain English what's wrong, I can try to fix it.

**Comment by GabenGar** [comment]
If you are not a typescript person then why do you provide typescript types instead of delegating them to [`DefinitelyTyped`](https://github.com/DefinitelyTyped/DefinitelyTyped)?
The problem at hand is not typescript per se but rather nodejs ESM/CJS shenanigans, and the [explanation of the problem](https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseExportDefault.md) is as plain english as it can get, considering the subject matter.
Your types say the package is a ESM module, when in reality it's CJS and the typescript need to know this distinction to work properly with nodejs ESM code. The distinction didn't matter back when ESM was just a mere syntactic sugar meant for feeding into a bundler and it just works™, but now that nodejs and even browsers support ESM, and both are pretty anal about the rules.

**Comment by arpowers** [comment]
I got it all working fine w "preserve" as module resolution mode. It does error due to CJS/ESM confusion if it's set to "NodeNext"...
